### ISSUE TITLE
4 elements returned back instead of with the json_query method

### ISSUE THREAD

**Issue by stdlef** [danielaparker/jsoncons#143]
I am experimenting with query_json method to retrieve some results from the following json

```
[
  {
    "root": {
      "id" : 10,
      "second": [
        {
             "names": [
            2
          ],
          "complex": [
            {
              "names": [
                1
              ],
              "panels": [
                {
                  "result": [
                    1
                  ]
                },
                {
                  "result": [
                    1,
                    2,
                    3,
                    4
                  ]
                },
                {
                  "result": [
                    1
                  ]
                }
              ]
            }
          ]
        }
      ]
    }
  },
  {
    "root": {
      "id" : 20,
      "second": [
        {
          "names": [
            2
          ],
          "complex": [
            {
              "names": [
                1
              ],
              "panels": [
                {
                  "result": [
                    1
                  ]
                },
                {
                  "result": [
                    1,
                    2,
                    3,
                    4
                  ]
                },
                {
                  "result": [
                    1
                  ]
                }
              ]
            }
          ]
        }
      ]
    }
  }
]
```

when I am quering with $..[?(@.result.length == 4)]

the result is:
```
[
    {
        "result": [1,2,3,4]
    },
    {
        "result": [1,2,3,4]
    },
    {
        "result": [1,2,3,4]
    },
    {
        "result": [1,2,3,4]
    }
]
```

returning 4 elements instead of two. I want to find 

- which element in the array has id 10 and result.length is 4

- which element in the array has an element in the "result" array with value 3 (both elements should returned)

Any ideas how to proceed with the query? 

Thank you in advance,
Lef

**Comment by danielaparker** [comment]
Returning 4 elements instead of 2 is a bug, thanks for identifying that. 

It can be reproduced with the simpler test case (returns 2 instead of 1):

```
    json j = json::parse(R"(
[
      {
        "result": [
          1,
          2,
          3,
          4
        ]
      }
]

)");

    json result = json_query(j,"$..[?(@.result.length == 4)]");
```

I think it's specific to how the selectors are handling the length of an array. I'll investigate.

Daniel

**Comment by danielaparker** [comment]
The issue with four elements returned is fixed on master. The issue had to do with how filters are applied to arrays, in the presence of recursion. In jsonpath, a filter appearing after an array is applied to each element of the array, which is effectively one level of recursion. If this is followed by recursion on the array, the filter is now applied to any object members in the array, which however have already been applied, hence the extra elements. I've introduced flags to prevent this.

To find which element in the array has id 10 and result.length is 4, use:
```
$..[?(@.id == 10)]..[?(@.result.length == 4)]
```
To find which element in the array has an element in the "result" array with value 3, use:
```
$..[?(@.result.length == 4)][?(@.result[0] == 3 || @.result[1] == 3 || @.result[2] == 3 || @.result[3] == 3)]
```
Note that filters can be chained, if you need recursion to get from one to the other, separate them with ".."

**Comment by danielaparker** [comment]
A related issue.

For good sense,
```
$..result[?(@.length == 4)]
```
should return
```
[[1,2,3,4],[1,2,3,4]]
```
however, that didn't return anything, because the filter was only being applied to the elements of the array, not the array itself. I've changed that to apply to the array as well, so that's fixed on master. 

I've included this version of your example in the documentation for the `json_query` function.

Daniel
